provider/cloudflare/cloudflare.go

```diff
@@ -342,6 +346,15 @@ func (p *CloudFlareProvider) Records(ctx context.Context) ([]*endpoint.Endpoint,
 func (p *CloudFlareProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
 	cloudflareChanges := []*cloudFlareChange{}
 
+	// if custom hostnames are enabled, deleting first allows to avoid conflicts with the new ones
+	if p.CustomHostnamesConfig.Enabled {
+		for _, endpoint := range changes.Delete {
+			for _, target := range endpoint.Targets {
+				cloudflareChanges = append(cloudflareChanges, p.newCloudFlareChange(cloudFlareDelete, endpoint, target, nil))
+			}
+		}
```

Comment:
Nested looping. Q(n*m), should be Q(n+m)

Reply:
This is the original code, with only added condition `if p.CustomHostnamesConfig.Enabled` - delete first if custom hostnames are enabled and keep deleting last (how it is now) otherwise.
We have `n` endpoints with `m` targets each, `n*m` is unavoidable as the current approach is "change per target"

Reply:
is avoidable, with a cost of a bit memory footprint.

```
first loop over deleted
  add targets to collection
end loop

second loop over targets
  do magic
end loop
```

Reply:
the targets are different for each endpoint, and there's [no intersection in `changes.Delete`, `changes.Create` and `changes.Update*`](https://github.com/kubernetes-sigs/external-dns/blob/master/plan/plan.go#L183)

with nested loops it's only scanning the same target exactly once or am i missing something?
<img width="1374" alt="Screenshot 2025-03-31 at 10 22 56 AM" src="https://github.com/user-attachments/assets/a4e018e1-55c5-4d6a-b20e-4e8bc7d90d8d" />

Reply:
I think this is very much on the edge. Typically (but not always) one loop nested in another could silently cause O(n²)

on line 351 `endpoint` collides with imported package name

Reply:
oh i see where the confusion is coming from
the `endpoint` package doesn't have the `Targets()` method, it's inner variable field
however, it's not great that we have that many conflicts in the code to be left to handle by go

for the consistence, I've renamed all such occurrences with variables named `e` when used internally, and `ep` for ) for function parameters `endpoint.Endpoint` type

none of it has been introduced with this PR, but it seems to be a nice occasion to make the code less ambiguous

thanks!

Reply:
ok